app/src/org/commcare/dalvik/activities/CommCareHomeActivity.java

```diff
@@ -210,9 +217,121 @@ private void processFromLoginLaunch() {
+                    CommCareApplication._().getCurrentApp().getAppPreferences()
+                            .getBoolean(CommCarePreferences.HAS_DISMISSED_PIN_CREATION, false);
+            if (!alreadyDismissedPinCreation || userManuallyEnteredPasswordMode) {
+                showPinChoiceDialog(loginMode);
+            }
         }
     }
 
+    private void showPinChoiceDialog(final LoginMode loginMode) {
+        String promptMessage;
+        if (CommCareApplication._().getRecordForCurrentUser().hasPinSet()) {
+            promptMessage = Localization.get("pin.dialog.prompt.reset");
+        } else {
+            promptMessage = Localization.get("pin.dialog.prompt.set");
+        }
+
+        final PaneledChoiceDialog dialog = new PaneledChoiceDialog(this, promptMessage);
+
+        DialogChoiceItem createPinChoice = new DialogChoiceItem(
+                Localization.get("pin.dialog.yes"), -1, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        dialog.dismiss();
+                        launchPinCreateScreen(loginMode);
+                    }
+                });
+
+        DialogChoiceItem nextTimeChoice = new DialogChoiceItem(
+                Localization.get("pin.dialog.not.now"), -1, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        dialog.dismiss();
+                    }
+                });
+
+        DialogChoiceItem notAgainChoice = new DialogChoiceItem(
+                Localization.get("pin.dialog.never"), -1, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        dialog.dismiss();
+                        CommCareApplication._().getCurrentApp().getAppPreferences()
+                                .edit()
+                                .putBoolean(CommCarePreferences.HAS_DISMISSED_PIN_CREATION, true)
+                                .commit();
+                        showPinFutureAccessDialog();
+                    }
+                });
+
+
+        dialog.setChoiceItems(new DialogChoiceItem[]{createPinChoice, nextTimeChoice, notAgainChoice});
+        dialog.addCollapsibleInfoPane(Localization.get("pin.dialog.extra.info"));
+        dialog.show();
+    }
+
+    private void showPinFutureAccessDialog() {
+        AlertDialogFactory f = AlertDialogFactory.getBasicAlertFactory(this,
+                Localization.get("pin.dialog.set.later.title"),
+                Localization.get("pin.dialog.set.later.message"), null);
+        f.showDialog();
+    }
+
+    private void launchPinAuthentication() {
+        Intent i = new Intent(this, PinAuthenticationActivity.class);
+        startActivityForResult(i, AUTHENTICATION_FOR_PIN);
+    }
+
+    private void launchPinCreateScreen(LoginMode loginMode) {
+        launchPinCreateScreen(loginMode, null);
```

Comment:
Instead of passing `null` in for `passwordFromAuth`, could you move the `getIntent().getStringExtra(LoginActivity.PASSWORD_FROM_LOGIN)` code up to here? 
Would remove the need for the `if (passwordFromAuth != null) {` line and avoid using `null` which is always a plus